### ISSUE TITLE
Primitive device: PNP/NPN (vertical oder lateral) for CMOS-only

### ISSUE THREAD

**Issue by hpretl** [IHP-GmbH/IHP-Open-PDK#267]
For CMOS-only designs, an NPN or PNP (either lateral or vertical) would be needed for bandgaps and similar uses.

**Comment by hpretl** [comment]
Thanks for the pointer, you are absolutely right! I totally missed it because there is no pcell in KLayout (at least none that I can find) and no mention in the process description, but there is a model and symbol for simulation and a fixed layout in `sg13g2_pr.gds`.

@KrzysztofHerman @sergeiandreyev Are there plans to provide a pcell, or is there one fixed size of PNP with a fixed layout (which is quite the norm for parasitic PNPs)? In any case, a description in the `SG13G2_process_spec.pdf` would be great!

**Comment by hpretl** [comment]
@sergeiandreyev cool! Do you have a rough timeline for the remaining cells?

**Comment by sergeiandreyev** [comment]
unfortunately, we don't have dedicated resources for this task, so the timeline is rather hard to define..
I agree, it will be useful for tracking, so I will create a separate GitHub 'enhancement' issue for new devices

**Comment by sergeiandreyev** [comment]
> In any case, a description in the `SG13G2_process_spec.pdf` would be great!

after checking this internally, the feedback is that we have a **new** `pnpMPA` device implementation in progress, so the process spec is planned to be updated with this **new** device

**Comment by hpretl** [comment]
> after checking this internally, the feedback is that we have a new pnpMPA device implementation in progress, so the process spec is planned to be updated with this new device

Excellent, good to hear! @sergeiandreyev Do you have a timeline for the bundle (doc, model, pcell)?

**Comment by sergeiandreyev** [comment]
> Do you have a timeline for the bundle (doc, model, pcell)?

around 6 month..

**Comment by olisnr** [comment]
because You allready give us the measurements of different pnpMPAs, it would be nice, You also give the test-chip-layouts as` .gds` with it. then the need for a PCell is much smaller. 
at the moment there is only the 1u/2u variant aviablle from the `ihp-sg13g2/libs.ref/sg13g2_pr/gds/sg13g2_pr.gds`

**Comment by olisnr** [comment]
im also interested in the SVaricap. is there more information about what it can and how to use?

**Comment by RTimothyEdwards** [comment]
@sergeiandreyev :  I have a follow-up question about the pnpMPA device.  Looking at the GDS of the device in the `sg13g2_pr` GDS (because there is no klayout pycell for it), there is nothing that uniquely identifies this device as a bipolar, which makes it impossible to uniquely identify the structure as a bipolar transistor for the purpose of netlist extraction.  The simple expedient of allowing the bipolar ID layer (TRANS.drawing) to cover the device would allow it to be extracted.  The PNP device would be differentiated from the SiGe devices by recognizing that the TRANS layer does not contain any emitter window.  Of course, since this would be a non-mask layer, you could just define any arbitrary unused layer type for a PNP device recognition layer.

**Comment by sergeiandreyev** [comment]
Hi @FaragElsayed2, could you please reply Tim how the `pnpMPA` recognition is done in KLayout LVS? I think we have this info in the documentation, but maybe it's better to provide more details..

**Comment by FaragElsayed2** [comment]
Hi @sergeiandreyev 

I assumed that the pnpMPA recognition is based on its geometry, the emitter, base, and collector layers are derived from their geometric arrangement, with the emitter island inside the base region, and the collector forming a surrounding ring, following the same structural defined in KLayout for lateral BJTs.

<img width="1306" height="499" alt="Image" src="https://github.com/user-attachments/assets/8cdc3e9f-841c-430e-ae19-4a5378651c7f" />

https://www.klayout.de/doc/manual/lvs_device_extractors.html

```
# =============
# ---- PNP ----
# =============

logger.info('Starting PNP-BJT DERIVATIONS')

pnp_mk = ptap_holes.not(pnp_exclude)

# pnp general nodes DERIVATIONS
pnp_e = pactiv.and(pnp_mk).and(nwell_iso)
pnp_b = nactiv.and(pnp_mk).and(nwell_iso)
pnp_c = ptap.interacting(pnp_mk).not(pnp_exclude)

# pnp_mpa nodes DERIVATIONS
pnp_mpa_e = pnp_e.and(pnp_b.extents).and(pnp_c.extents)
pnp_mpa_b = pnp_b.interacting(pnp_b.extents.interacting(pnp_mpa_e))
pnp_mpa_c = pnp_c.interacting(pnp_c.extents.interacting(pnp_mpa_e))
```

https://github.com/IHP-GmbH/IHP-Open-PDK/blob/62c1d640dc1c91f57bc1a8e4e08e537a7a105ae8/ihp-sg13g2/libs.tech/klayout/tech/lvs/rule_decks/bjt_derivations.lvs#L110

**Comment by RTimothyEdwards** [comment]
@FaragElsayed2 :  That is basically what I am doing to identify a PNP device.  Thanks for the confirmation.

But stil, it seems rather arbitrary, easy to think of geometry that would be identified as a PNP but which is not a PNP, and it would be trivially easy to add an identifier layer for PNPs (including just using the existing TRANS marker, as I indicated above) and would make the extraction process both less complicated, and faster, and eliminate any chance that layout might be interpreted incorrectly.  Just my two cents (and noting that yesterday, the U.S. stopped minting pennies, so my two cents is now valueless).

**Comment by FaragElsayed2** [comment]
@RTimothyEdwards 

I agree with that. Other technologies do this, I believe it’s an IHP PDK team decision.

**Comment by sergeiandreyev** [comment]
ok, I'll discuss this internally

**Comment by RTimothyEdwards** [comment]
@ahmad2002-hub :  The klayout pycell and magic gencell for the PNP device both work for SG13G2.  The discussion above was only about implementation issues, as I consider the device recognition method to be "fragile".  But it basically works.

There is an entirely different issue about the use of the pnpMPA device in SG13CMOS5L, because the drawn device should behave "more or less" the same in both processes, but CMOS5L lacks a deep nwell, so the behavior will be somewhat different, and there is not yet any SPICE model for the device in CMOS5L.  I used the pnpMPA for a bandgap circuit here:  https://github.com/RTimothyEdwards/sg13cmos5l_ocd_ip__biasgen/.  But that was with the understanding that the device model is unknown and therefore the bandgap cannot be fine-tuned on the first attempt.  I expect it to behave like a poorly-calibrated bandgap.
